[웹개발 3주차, 4주차 미션] 구현 완료했습니다. 리뷰 부탁드립니다.

sunghyuki opened the pull request:
## 3주차 미션에서의 의문

JPA를 공부하며 계층을 Controller-Service-Repository로 구성한 후 JPA의 save 메서드를 이용해 데이터를 저장하면 데이터는 영속성 컨텍스트에 관리되는 "영속" 상태가 된다. 영속성 컨텍스트의 변경 감지(Dirty Checking) 기능으로 데이터에 변경이 생긴다면 1차 캐시의 스냅샷과 비교해 트랜잭션 커밋 시 영속성 컨텍스트의 변경 내용을 데이터베이스에 반영한다.

하지만 이번 미션에서 데이터를 update 시켰지만 데이터의 변경이 일어나지 않았다. 왜일까?
<br>

## 트랜잭션이란?

* 데이터베이스의 상태를 변화시키기 위해서 수행하는 작업의 단위
* 트랜잭션의 성질 (ACID)
  * 원자성(Atomicity)
    * 하나의 트랜잭션이 더 작게 나눌 수 없는 최소의 단위라는 뜻
  * 일관성(Consistency)
    * 트랜잭션이 완료된 결괏값이 일관적인 데이터베이스 상태를 유지하는 것
  * 독립성(Isolation)
    * 트랜잭션이 수행되고 있을 때, 다른 트랜잭션의 연산 작업이 중간에 끼어들어 기존 작업에 영향을 주지 못하도록 하는 것
  * 지속성(Durability)
    * 트랜잭션을 성공한 후 데이터베이스에 반영된 것은 영원히 반영되어야 한다는 것
<br>

어떤 사람과 거래를 하고 있는 상황이다.

- 물건을 받기 위해 상대방에게 선입금을 했다.
- 그런데 벽돌이 도착했다.

사기당한 사실을 깨닫고 나면, 시간을 되돌리고 싶지 않을까?

트랜잭션 개념은 이를 가능하게 한다.

상대방에게 입금하는 작업이 성공했더라도 원하던 물건을 받지 못했다면 입금이 취소되었을 것이다.

모든 작업들이 성공적으로 완료되어야 작업 묶음의 결과를 적용하고, 어떤 작업에서 오류가 발생했을 때는 이전에 있던 모든 작업들이 성공적이었더라도 없었던 일처럼 완전히 되돌리는 것이 트랜잭션의 개념이다.

데이터베이스를 다룰 때 트랜잭션을 적용하면 데이터 추가, 갱신, 삭제 등으로 이루어진 작업을 처리하던 중 오류가 발생했을 때 **모든 작업들을 원상태로 되돌릴 수 있다**. 모든 작업들이 성공해야만 최종적으로 데이터베이스에 반영하도록 한다.
<br>

## 롤백과 커밋

* 롤백
  * 트랜잭션의 원자성을 위해 데이터 변경사항이 취소되면 데이터의 이전 상태로 복구
* 커밋
  * 하나의 트랜잭션 작업이 정상적으로 끝마쳐졌을 때 디비에 변경사항 반영
<br>

## 스프링에서 `@Transactional` 을 이용하여 트랜잭션 처리를 하는 방법

DB와 관련된, 트랜잭션이 필요한 서비스 클래스 혹은 메서드에 `@Transactional` 어노테이션을 달아주면된다.

클래스, 메서드 모두에 `@Transactional` 어노테이션을 붙이면 메서드 레벨의 `@Transactional` 선언이 우선 적용된다. (선언적 트랜잭션)

```java
//게시판의 게시글을 저장하는 메서드 
@Transactional 
public void save(Question question) { 		
  questionRepository.save(queestion); // 게시글 저장
}
```

`@Transactional`이 붙은 메서드는 메서드가 포함하고 있는 작업 중에 하나라도 실패할 경우 전체 작업을 취소한다.
<br>

* transaction `begin`, `commit`을 자동 수행해준다.

* 예외를 발생시키면, `rollback` 처리를 자동 수행해준다.
<br>

이유 : Transaction commit이 발생하지 않았기 때문에 update 시 데이터베이스에 반영되지 않는다.
<br>

* Repository 계층의 각 메서드 레벨에 선언적 트랜잭션이 적용되어 있다.

```java
@Transactional
public <S extends T> S save(S entity) {

    if (entityInformation.isNew(entity) {
        em.persist(entity);
        return entity;
    } else {
        return em.merge(entity);
    }
}
```
<br>

@Transactional 어노테이션을 붙이지 않았기 때문에 try-catch를 이용해 구현한 경우 update 메서드를 이용해서 회원 정보 update 가능.

```java
@Autowired
PlatformTransactionManager platformTransactionManager;
```

* PlatfromTransactionManager
  * 스프링 트랜잭션 추상화의 핵심 인터페이스


```java
@PostMapping("/users/{id}/update")
public String update(@Pathvariable Long id, User updatedUser) {
    DefaultTransactionDefinition defaultTransactionDefinition = new DefaultTransactionDefinition();
    TransactionStatus transactionStatus = platformTransactionManager.getTransaction(defaultTransactionDefinition);

    User savedUser = userRepository.findById(id).orElseThrow(RuntimeException::new);

    try {
        if (savedUser.equalsPassword(updatedUser.getPassword()) {
            savedUser.update(updatedUser);
            platformTransactionManager.commit(transactionStatus); // 트랜잭션 매니저를 통해 직접 커밋
        }
    } catch (Exception exception) {
        platformTransactionManager.rollback(transactionStatus); // 트랜잭션 매니저를 통해 직접 롤백
    }

    return "redirect:/users";
}
```
<br>

글쓴이를 disabled 옵션을 이용해 비활성화시킨 상태에서 Post를 보내면 글쓴이는 org.h2.jdbc.JdbcSQLException: NULL not allowed for column "WRITER"; SQL statement:

-> 글쓴이 수정을 막기 위해 input 태그 옵션으로 disabled를 사용하였지만 disabled 옵션은 비활성화와 함께 form으로 전송되지 않는다. 입력은 비활성화하고 폼으로 전송 가능하게 하기 위해 readonly 옵션 추가

sunghyuki committed:
refactor : 피드백 수정
- Question 도메인 index 동일성 판단 메서드 추가
- User 도메인 password 동일성 판단 메서드 수정
- User 도메인 동일 사용자 판단 메서드 추가

sunghyuki committed:
chore : H2 DB Connection 설정
- H2 연결 설정
- H2 DB 콘솔 접근 설정
- 실행 쿼리 보기 설정
- 테이블 자동 생성 설정

sunghyuki committed:
refactor : User 클래스를 DB 테이블에 매핑

sunghyuki committed:
feat : UserRepository 인터페이스 추가

sunghyuki committed:
refactor : 사용자 데이터 추가 비즈니스 로직 수정
- UserRepository의 save() 메서드를 통해 DB에 사용자 데이터 추가

sunghyuki committed:
refactor : 사용자 목록 조회 비즈니스 로직 수정
- UserRepository의 findAll() 메서드를 통해 DB에 저장되어 있는 전체 사용자 목록 조회

sunghyuki committed:
refactor : 사용자 조회 비즈니스 로직 수정
- UserRepository의 findOne() 메서드를 통해 DB에 저장되어 있는 사용자 데이터 조회

sunghyuki committed:
feat : QuestionRepository 인터페이스 추가

sunghyuki committed:
refactor : Question 클래스를 DB 테이블에 매핑

sunghyuki committed:
refactor : 질문 데이터 저장 비즈니스 로직 수정
- QuestionRepository의 save() 메서드를 통해 DB에 질문 데이터 추가

sunghyuki committed:
refactor : 질문 목록 조회 비즈니스 로직 수정
- QuestionRepository의 findAll() 메서드를 통해 DB에 저장되어 있는 전체 질문 목록 조회

sunghyuki committed:
refactor : 질문 상세보기 비즈니스 로직 수정
- QuestionRepository의 findById() 메서드를 통해 DB에 저장되어 있는 질문 데이터 조회

sunghyuki committed:
refactor : 회원정보 수정 비즈니스 로직 수정
- UserRepository의 findById() 메서드를 통해 DB에 저장되어 있는 사용자 정보 불러오기

sunghyuki committed:
refactor : 회원정보 수정 비즈니스 로직 수정
- UserRepository의 findById() 메서드를 통해 DB에 저장되어 있는 사용자 정보를 불러오고 사용자 정보 업데이트 후 저장

sunghyuki committed:
refactor : Optional 값 처리

sunghyuki committed:
refactor : 질문 수정 비즈니스 로직 수정
- textarea 태그 밖에 value 값을 넣어주면 내용이 보인다.

sunghyuki committed:
feat : 질문 삭제 기능 추가

hyukjin-lee reviewed (approved):
고생하셨습니다 ~
코드가 거의 같으니 진영님의 리뷰를 함께 보시면 좋을 것 같습니다 ~
이 PR 에서는 개선된 부분도 보이는 것 같은데 서로 페어하면서 협의가 된건지도 궁금하네요

sunghyuki committed:
fix : 질문 관련 API 버그 해결
- HTML a 태그 링크 수정
- 글쓴이 수정을 막기 위해 input 태그 옵션으로 disabled를 사용하였지만 disabled 옵션은 비활성화와 함께 form으로 전송되지 않는다. 입력은 비활성화하고 폼으로 전송 가능하게 하기 위해 readonly 옵션 추가
- qnaForm GET 요청 메서드 추가
- 질문 수정 기능 추가

sunghyuki commented:
> 고생하셨습니다 ~ 코드가 거의 같으니 진영님의 리뷰를 함께 보시면 좋을 것 같습니다 ~ 이 PR 에서는 개선된 부분도 보이는 것 같은데 서로 페어하면서 협의가 된건지도 궁금하네요

옙 함께 페어하면서 협의했습니다!

sunghyuki committed:
fix : 회원 관련 API 버그 해결
- HTML a 태그 링크 수정
- 회원 userId 수정을 막기 위해 input 태그 옵션으로 disabled를 사용하였지만 disabled 옵션은 비활성화와 함께 form으로 전송되지 않는다. 입력은 비활성화하고 폼으로 전송 가능하게 하기 위해 readonly 옵션 추가
- 회원가입 폼 GET 요청 메서드 추가

sunghyuki committed:
refactor : 쓰지 않는 파라미터 제거

sunghyuki committed:
refactor : Optional의 사용 목적에 수정
- DB로부터 데이터를 조회할 때 데이터가 존재하지 않을 경우 예외를 던져주고 데이터가 존재할 경우만 다음 라인으로 넘어갈 수 있게 수정

sunghyuki committed:
feat : SLF4J를 이용한 로깅 추가
- 롬봄 의존성 추가
- 유저 생성 시 유저 객체를 로그를 통해 확인

sunghyuki committed:
refactor : Repository 인터페이스 변경

sunghyuki committed:
feat : UserRepository에서 userId에 해당하는 데이터 조회 메서드 추가
- JpaRepository 상속

sunghyuki committed:
fix : HTML 중복 제거 문제 해결
- updateForm만 레이아웃 방식으로 중복을 제거해주는 기능이 적용이 안 되었는데 해결

sunghyuki committed:
feat : 로그인 기능 구현
- 로그인 비즈니스 로직 구현
- mustache 세션 설정 추가
- mustache 로그인 세션 데이터를 사용한 if/else 구문 추가

sunghyuki committed:
feat : 개인정보 수정 폼 반환 로직 수정
- 세션 ID가 null인 경우 login 페이지로 redirect
- 세션 ID가 null이 아닌 경우 updateForm 반환
- 테스트 데이터 추가

sunghyuki committed:
feat : 개인정보 수정 비즈니스 로직 세션 정보 추가

sunghyuki committed:
refactor : 개인정보 업데이트 로직 수정
- 세션으로부터 받아온 Object 타입의 유저 정보를 받아옴과 동시에 형변환
- 질문하기 버튼 클릭 시 로그인된 사용자라면 질문하기 페이지로, 로그인이 안 된 사용자라면 로그인 페이지로 redirect
- qna/form.html 글쓴이 입력 필드 삭제

sunghyuki committed:
feat : 질문 도메인과 회원 도메인을 ManyToOne 연관관계로 매핑

sunghyuki committed:
refactor : 개인정보 업데이트 폼 비즈니스 로직 수정
- 세션으로부터 받아온 Object 타입의 유저 정보를 받아옴과 동시에 형변환

sunghyuki committed:
feat : 회원 도메인과 질문 도메인 연관관계 매핑을 위해 QuestionRepository에 JpaRepository 상속

sunghyuki committed:
feat : 로그인한 사용자만 질문하기 화면에 접근 권한 기능 추가

sunghyuki committed:
feat : 로그인한 사용자만 질문 상세보기 페이지에서 수정 삭제 기능 추가
- 현재의 요구사항보다 사용자의 입장에서 로그인된 사용자에게만 수정 삭제 버튼이 보이는 것이 맞다고 판단하여 요구사항 수정 구현

sunghyuki committed:
feat : 로그인한 사용자만 질문 수정 가능하도록 기능 추가

sunghyuki committed:
feat : 로그인한 사용자만 질문 삭제 가능하도록 기능 추가

Bellroute commented on src/main/java/codesquad/domain/question/Question.java:
```diff
@@ -0,0 +1,63 @@
+    private User writer;
+
+    @Column(nullable = false, length = 200)
+    private String title;
```
작성하신 로직대로면 title의 길이 제한이 데이터베이스 수준에서 체크가 될 것 같습니다. 이전 단계에서 데이터를 검증하는 로직이 추가된다면 데이터베이스 수준에서 체크하는 것과 어떤 차이가 있을까요?

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
+        User sessionedUser = (User) httpSession.getAttribute("sessionedUser");
+        Question savedQuestion = questionRepository.findById(index).orElseThrow(NoSuchElementException::new);
+
+        if(!sessionedUser.isSameUser(savedQuestion.getWriter().getUserId())) {
```
혹시 연속되는 getter 사용이 언짢지는 않으셨나요? 동일 유저인지 판단을 Question 객체가 직접한다면 코드가 어떻게 달라질 수 있을까요?

kanamycine replied:
감사합니다. 언짢아서 말씀해주신대로 처리하였습니다 !!

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
-        question.setIndex(index++);
-        questions.add(question);
+    public String create(Question question, HttpSession httpSession) {
+        User sessionedUser = (User) httpSession.getAttribute("sessionedUser");
```
1. "sessionedUser"라는 key를 다른 문자열로 바꿔야 하는 일이 발생한다면 코드의 여러군데를 수정해야할 것 같습니다. 이러한 경우를 고려해 리팩토링해볼 수 있을 것 같습니다.
2. (현우님께도 드렸던 내용이지만) HttpSession은 동일한 key("sessionedUser")로 유저 정보를 불러오는데, 어떻게 사용자마다 그에 맞는 정보를 불러오는 걸까요? html 쪽을 작성하면서 question 정보와는 다르게 HttpSession 정보는 넘겨준적이 없는데 어디서 추가된 걸까요?

kanamycine replied:
2. 

**의문점**

```java
HttpSession.getAttribute("user");

사용자 A가 접속해도 "user" Key로 값을 가져오고,
사용자 B가 접속해도 "user" Key로 가져온다. 
같은 Key를 쓰는데 어떻게 유저 A와 B를 구분해서 값을 가져오나?
```

 이것이 가능한 이유를 알기 위해서는 JSESSIONID라는 것을 알아야 합니다.

**JSESSIONID ?**

- 톰캣 컨테이너에서 세션을 유지하기 위해 발급하는 키
- 상태를 저장하기 위해서 톰캣은 JSESSIONID 쿠키를 클라이언트(브라우저)에게 발급해주고 이 값을 통해 세션을 유지할 수 있도록 한다.

> The name to be used for all session cookies created for this context. If set, this overrides any name set by the web application. If not set, the value specified by the web application, if any, will be used, or the name JSESSIONID if the web application does not explicitly set one.
>
> 요약 : 이름을 지정하지 않으면 JSESSIONID 이름이 사용됩니다. 

기존 세션을 사용하고 있는 URL을 Postman을로 API 호출해보았습니다.

![image-20220406160119421](/Users/samuel/Library/Application Support/typora-user-images/image-20220406160119421.png)

그러면 Response Cookie에 JSESSIONID가 포함되어 응답으로 오는 것을 볼 수 있습니다. 즉, 세션이 서버에 등록되면 서버에서 응답값으로 클라이언트에게 JSESSIONID를 RESPONSE HEADER에 담아 보내주는 것을 볼 수가 있습니다.

즉, 전체적인 흐름을 정리하면 아래와 같습니다.

1. 클라이언트(웹 브라우저의 사용자)가 처음으로 웹 어플리케이션을 방문하거나 request.getSession()을 통해 HttpSession을 처음으로 가져 오면 `서블릿 컨테이너`는 새로운 HttpSession 객체를 생성하고 길고 unique한 ID를 생성 후, 서버의 메모리에 저장합니다.
2. 서블릿 컨테이너는 `JSESSIONID`란 이름을 key로, 생성한 session ID를 value로 하여 HTTP 응답의 Set-Cookie header에 cookie로 설정합니다
3. 브라우저는 다음 요청부터 `Request Headers`에 `JSESSIONID`를 담아서 서버로 요청을 보냅니다.
4. `서블릿 컨테이너`는 들어오는 모든 HTTP request의 cookie header에서 `JSESSIONID`라는 이름의 cookie가 있는지 확인하고 해당 값 (session ID)을 사용하여 서버의 메모리에 저장된 HttpSession을 가져옵니다

그러면 지금까지 위에서 정리한 내용을 바탕으로 `같은 Key를 쓰는데 어떻게 유저 A와 B를 구분해서 값을 가져오나요?` 라는 질문은 요청이 올 때의 `Request Header에 존재하는 JSESSIONID로 구분할 수 있다.` 라고 답 할 수 있겠습니다.  

`setAttribute` 할 때의 `key`가 중요한 것이 아니라 클라이언트(브라우저)에서 요청을 보낼 떄 `Request Header`에 `JESSIONID`가 있냐 없냐가 유저를 구분하는데 중요한 역할을 한다는 것을 알 수 있습니다.

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
+        return mav;
+    }
+
+    @GetMapping("/questions/{index}/updateForm")
```
REST API에서 권장하는 url 규칙에는 소문자만 사용한다는 규칙이 있습니다. 그 이유는 무엇일까요?(힌트 - 웹지탱 5장)

kanamycine replied:
**URI 경로는 소문자가 적합합니다**

URI 경로에 대문자 사용은 피하도록 해야 합니다. 대소문자에 따라 다른 리소스로 인식하게 되기 때문입니다. RFC 3986(URI 문법 형식)은 URI 스키마와 호스트를 제외하고는 대소문자를 구별하도록 규정하기 때문입니다.

또 한가지의 이유는 특정 언어의 컨벤션에 의존하는 문자열을  URI에 포함시키면 안되기 때문입니다 ! !

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
+
+    @PutMapping("/questions/{index}")
+    public String update(@PathVariable Long index, Question updatedQuestion, HttpSession httpSession) {
+        User sessionedUser = (User) httpSession.getAttribute("sessionedUser");
```
해당 요청에 대해서는 세션 정보를 이용한 권한 체크 로직이 없는 것 같습니다. 누락된 건가요??

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
+    public String update(@PathVariable Long index, Question updatedQuestion, HttpSession httpSession) {
+        User sessionedUser = (User) httpSession.getAttribute("sessionedUser");
+        Question savedQuestion = questionRepository.findById(index).orElseThrow(NoSuchElementException::new);
+        if(savedQuestion.equalsIndex(index)) {
```
savedQuestion이 null이 아닌 이상 이 로직의 결과가 false인 경우가 있을까요?

sunghyuki replied:
null이 존재하지 않는 이상 이 로직의 의미가 없는 것 같아 보입니다. 로직을 다르게 구성할 필요가 있어 보이네요 ~ 질문 도메인의 writer 정보와 세션으로부터 가져온 유저 정보를 비교하여 질문 업데이트 가능하도록 하면 될 것 같습니다. 감사합니다

Bellroute commented on src/main/java/codesquad/web/QuestionController.java:
```diff
@@ -1,42 +1,87 @@
+    @DeleteMapping("/questions/{index}")
+    public String delete(@PathVariable Long index, HttpSession httpSession) {
+        User sessionedUser = (User) httpSession.getAttribute("sessionedUser");
+        Question question = questionRepository.findById(index).orElseThrow(NoSuchElementException::new);
```
NoSuchElementException이 발생하면 어떤 페이지가 뜨게 되나요? 그리고 http 응답코드는 몇번으로 넘어가나요?

kanamycine replied:
500이 뜨고, 이는 예외적인 또는 예측하지 못한 에러의 경우 500에러가 출력된다고 생각할 수 있습니다.

- 서버 통신의 Timeout 시간 지연 오류
- 서버 트래픽 과부하
- 서버 언어의 구문 에러 (스크립트 문법 오류)

분류되지 못한 오류 신호로 자세한 정보를 담고 있지는 않기 때문에 정확한 원인을 파악하기 위해서는 정보가 부족한게 사실입니다. 특히 서버의 로그 정보를 확인하여 원인을 파악하는 것이 나을 수 있겠습니다. 


현재는 NoSuchElementException이 터졌을 때 따로 리다이렉트 되는 html을 구현하지 않아서 그대로 에러 페이지가 노출되게 되어있습니다.

Bellroute commented on src/main/java/codesquad/web/UserController.java:
```diff
@@ -1,65 +1,95 @@
-                .filter(user -> user.getUserId().equals(userId))
-                .findFirst()
-                .get();
+    @PostMapping("/users/{id}/update")
```
1. 해당 요청의 http 메소드를 POST로 하신 이유가 있을까요?
2. RESTful한 웹 서비스를 고려한다면 자원에 대한 행위는 HTTP 메소드로 표현하고, 자원을 나타내는 uri 자체에는 행위(update)를 포함시키지 않는게 좋을 것 같습니다.

Bellroute commented on src/main/java/codesquad/web/UserController.java:
```diff
@@ -1,65 +1,95 @@
 
-        if(savedUser.validatePassword(updatedUser)) {
-            savedUser.update(updatedUser);
+        if (sessionedUser.equalsPassword(updatedUser.getPassword())) {
```
비밀번호만 비교해도 상관없을까요? sessionUser와 updawtedUser는 다른 사용자인데 우연히 비밀번호가 동일하다면 어떻게 될까요?

sunghyuki replied:
현실 세계를 투영해 봤을 때 다른 사용자끼리 비밀번호가 같은 경우가 존재할 수 있으니 클라이언트에서 넘어온 url의 id값과 세션의 id값을 비교해서 해결할 수 있을 것 같네요~

Bellroute reviewed (commented):
수고 많으셨습니다! 현우님 pr리뷰에서 말씀드린 것과 동일하게 Session에 대해서 학습해보시는 걸 추천합니다bb
추가로 lombok에 대한 의존성을 추가하셔서 사용하셨는데 lombok에서 제공하는 기능들+사용시 주의할 점에 대해 학습해보시는 것도 좋을 것 같습니다!
